sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuJsonTuple.scala

```diff
@@ -50,59 +49,58 @@ case class GpuJsonTuple(children: Seq[Expression]) extends GpuGenerator
     }
   }
 
-  def generate(inputBatch: ColumnarBatch,
+  override def generate(inputBatch: ColumnarBatch,
       generatorOffset: Int,
       outer: Boolean): ColumnarBatch = {
-
-    val jsonGpuExpr = GpuBindReferences.bindReference(jsonExpr, 
-      Seq(jsonExpr.asInstanceOf[NamedExpression].toAttribute))
+    
+    val json = inputBatch.column(generatorOffset).asInstanceOf[GpuColumnVector].getBase
     val schema = Array.fill[DataType](fieldExpressions.length)(StringType)
+    val nonGeneratorCols = (0 until generatorOffset).map { i =>
+      inputBatch.column(i).asInstanceOf[GpuColumnVector].incRefCount
+    }
 
-    withResource(columnarEvalToColumn(jsonGpuExpr, inputBatch).getBase) { json =>
-      val fieldScalars = fieldExpressions.safeMap { field =>
-        withResourceIfAllowed(field.columnarEval(inputBatch)) { fieldVal =>
-          fieldVal match {
-            case fieldScalar: GpuScalar =>
-              // Specials characters like '.', '[', ']' are not supported in field names
-              Scalar.fromString("$." + fieldScalar.getBase.getJavaString)
-            case _ => throw new UnsupportedOperationException(s"JSON field must be a scalar value")
-          }
+    val fieldScalars = fieldExpressions.safeMap { field =>
+      withResourceIfAllowed(field.columnarEval(inputBatch)) { fieldVal =>
+        fieldVal match {
+          case fieldScalar: GpuScalar =>
+            // Specials characters like '.', '[', ']' are not supported in field names
+            Scalar.fromString("$." + fieldScalar.getBase.getJavaString)
+          case _ => throw new UnsupportedOperationException(s"JSON field must be a scalar value")
         }
       }
-      withResource(fieldScalars) { fieldScalars =>
-        withResource(fieldScalars.safeMap(field => json.getJSONObject(field))) { resColumns =>
-          withResource(new Table(resColumns: _*)) { tbl =>
-            GpuColumnVector.from(tbl, schema)
-          }
+    }
+
+    withResource(fieldScalars) { fieldScalars =>
+      withResource(fieldScalars.safeMap(field => json.getJSONObject(field))) { resultCols =>
+        val generatorCols = resultCols.map(_.incRefCount).zip(schema).map { case (col, dataType) =>
```

Comment:
This needs to be a `safeMap` otherwise if `GpuColumnVector.from` throws then we'll leak. the columns that have already been iterated.

Reply:
To be clean you need a safeMap for both of the map calls.

Reply:
Updated. Thank you!